### ISSUE TITLE
Does mergerfs pass through io priority?

### ISSUE THREAD

**Issue by kotarou3** [trapexit/mergerfs#932]
Does mergerfs pass through io priority?

Imagine I have two processes A and B, where process A doesn't use the disk much but is latency sensitive, while process B can completely saturate the disk while latency is unimportant.  
This means that I would assign process A a higher io priority (lower ionice) than B. For example, A = 0, B = 7.

If both processes access the same underlying disk via mergerfs, is the io priority of each access preserved?

I'm asking this since I'm observing vastly increased latency on process A while process B is hammering the disk even though I've set process B to a lower priority, so I'm wondering if this is a mergerfs problem or something deeper

**Comment by trapexit** [comment]
I'm not sure how ioprio_set interacts with FUSE. It's certainly not something that FUSE articulates in it's protocol and therefore not expressly actioned in any way by mergerfs. And given mergerfs is its own userland process wouldn't be generally impacted by any priority outside thread that perhaps would get caught up in IOPRIO_WHO_USER calls. I suppose mergerfs could look up the value using ioprio_get and then set the active thread's priority to match but I'm not sure that would be a good idea as it would be constantly changing and add 2 extra syscalls to every function. And i'm not sure if that would confuse the scheduler with that many priority changes.

io priority applies to users, process groups and process or threads. mergerfs doesn't have pools of threads per uid/gid. Arguably it should try to do so but it isn't possible right now. If it did work that way mimicking the client pid's priority would be more practical.

**Comment by kotarou3** [comment]
Perhaps a "temporary" solution could be to:
 1. Have a pool of "default" priority threads (like right now), and a low priority thread
 2. On a request, look up the io priority of the thread
 3. If it's lower than the "default" priority, pass the request to the low priority thread

That way higher priority requests will only incur 1 extra syscall, and there won't be frequent changing of priority

This could all be opt-in with a mount option, so people who don't care about priority won't incur any performance hit at all

**Comment by trapexit** [comment]
That'd still require a decent rewrite of the thread management. Which right now is pretty simple. At the point the thread workload is dispatched the message isn't even decoded. This is why this kind of message level change is pretty invasive. Threads wait on the fuse character device for messages and then just work on them. The whole thing needs to be rewritten so that there are reader threads and then workers where the readers do some basic message parsing to find out the info they need to dispatch. 

mergerfs sets its own priority to -10 but it is not something that can be changed from the command line. You could compile mergerfs yourself and change it to see if that helps some. If it did I could make it a configurable option.

**Comment by Jiinja** [comment]
I'm in a very similar situation as kotarou3 and can't find any information on how io priority works with mergerfs. I see that you can configure CPU scheduling priority, but I don't see anything about io priority. Copying a file at idle priority using rsync and ionice from a separate drive to a mergerfs pool, I can see with iotop that the reads on the first drive are idle priority but the writes are via mergerfs at the io priority of mergerfs. Is there any way to use ionice with mergerfs? Does the passthrough option pass-through the io priority to the disk? Can I use two mergerfs pools for the same drives, one with normal priority and one with less? 

The problem I am trying to solve is plex stuttering while copying large remux files (50-150gb) into mergerfs pool. Currently, I have ionice + rsync with a low bandwidth limit (20MB/s) that has improved the situation, but now file transfers take ages when they don't necessarily need to. I was checking iotop during a transfer and noticed that it doesn't seem like ionice is respected by mergerfs. I'd much prefer not limiting the bandwidth of the transfer/copy and instead relying on iopriority to serve plex what it wants when it needs and copy the files quickly when it can. Let me know if I'm off-base or if I need to make a new issue.

**Comment by trapexit** [comment]
> but I don't see anything about io priority. 

Because there simply is nothing in mergerfs related to io priority. 

FUSE offers no way to pass the scheduler and priority details of the request being made. The best that could be done is querying the thread/process making the request's details every request (which would be rather expensive) or on of the opening of a file and cached for the lifetime of the open.

Since the majority of users don't change io prio values of their software it hasn't really been a concern.

>  Is there any way to use ionice with mergerfs?

Sure. mergerfs is just another piece of userland software. Since it doesn't set any priorities there is nothing it could overwrite (currently.)

> Does the passthrough option pass-through the io priority to the disk?

Passthrough isn't available yet but no... mergerfs is its own process with its own priorities, limits, etc. At best mergerfs can do is what is described above.

I can look into adding ioprio_get and set on file open for a future release but it'll need to be optional (assuming it works as expected) since that will add additional overhead.

While not fine grained as described above note, from the ioprio man page:

> If no I/O scheduler has been set for a thread, then by default the I/O priority will follow the CPU nice value ([setpriority(2)](https://man7.org/linux/man-pages/man2/setpriority.2.html)).

So maybe not what you're looking for but you do have the ability to change the CPU priority currently.

**Comment by trapexit** [comment]
I'll prototype something for the next release. If I decide to kick it to the next next release I'll update the docs about the topic for the time being.

**Comment by trapexit** [comment]
Also... keep in mind that not all schedulers support priority. If your device shows a scheduler of "none" or a few others the priority value has no meaning.

**Comment by Jiinja** [comment]
I made a duplicate pool with cpu priority 10 for the imports and that seems to be working for now. Thanks for the tip

**Comment by trapexit** [comment]
I prototyped something and at least mechanically it is functional. IE... reads the io prio of the thread making the request and sets the mergerfs thread to that value when doing reads, writes, and directory reads. I can provide you a static build to test if you like. 

However, I've learned that with passthrough mergerfs shouldn't have to do anything else as the IO should be happening in the context of the calling process/thread.

**Comment by Jiinja** [comment]
I'm down to test that out, would be nice to just have one pool. As for passthrough I'm not too concerned about the functionality I'd lose when using it eventually. Are the performance concerns and/or buffer bloat with page caching effectively negligible with the performance increase of passthrough?

**Comment by trapexit** [comment]
Passthrough is really passthrough. mergerfs is removed from the read/write path. So none of those issues, in relation to mergerfs, are relevant any more.

This is a build with ioprio. It's not configurable. This is based on the latest codebase so passthrough is available for you to test as well.

[mergerfs-static-linux_amd64.tar.gz](https://github.com/user-attachments/files/22049096/mergerfs-static-linux_amd64.tar.gz)

**Comment by Jiinja** [comment]
Finally got some time to try this out and it seems to be working. I'm not seeing stuttering in plex when large files are copied and I can see the priority I set using ionice reflected in the iotop output for mergerfs when a file transfer is in progress.

A separate issue I was running into when using two mergerfs pools was plex not detecting file changes/additions. I assume this was because adding files using another pool is "out of band." Going back to one pool seems to have resolved this issue as well. Thanks for putting this build together!

I am currently running proxmox 8 running kernel 6.8 so I can't test passthrough.

**Comment by trapexit** [comment]
>  I assume this was because adding files using another pool is "out of band."

https://trapexit.github.io/mergerfs/latest/faq/compatibility_and_integration/#does-inotify-and-fanotify-work

Yes, it's not possible for the kernel to know changes are made if they are made out of band.

Regarding the priority setting. I'll see if I can turn the feature into something more configurable and production ready but might need to wait till after 2.41.0 gets out. We'll see.
